### ISSUE TITLE
[css-values] Use of 100vw is causing pointless horizontal scrollbars on some websites

### ISSUE THREAD

**Issue by simevidas** [w3c/csswg-drafts#6026]
Windows users and macOS users that have enabled “Show scroll bars: Always” are seeing pointless horizontal scrollbars on some websites that use the `100vw` value. One common use-case for `100vw` are full-width elements:

```css
.full-width {
    width: 100vw;
    position: relative;
    left: 50%;
    right: 50%;
    margin-left: -50vw;
    margin-right: -50vw;
}
```

Examples:

* https://timkadlec.com/remembers/2021-01-26-what-about-ember/
* https://blog.maximeheckel.com/posts/the-power-of-composition-with-css-variables
* https://fugu-tracker.web.app/
* https://ourworldindata.org/coronavirus/country/germany (**[screenshot](https://twitter.com/simevidas/status/1393693084173623297)**)
* https://www.monotype.com/fonts/helvetica-now-variable (**[screenshot](https://twitter.com/simevidas/status/1427675694822199299)**)

If you can’t check for yourself and are not sure what a pointless horizontal scrollbar is, watch this [video that I posted on Twitter](https://twitter.com/simevidas/status/1363019736896921604).

The authors of these websites are most likely not aware of this issue because they developed the website on macOS with default (overlaid) scrollbars, where this issue does not occur.

As a result, many people are experiencing these pointless horizontal scrollbars, and a fair share of these people probably find them annoying (*raises hand*  🙋‍♂️). 

Does the Working Group consider this to be a problem that requires an intervention? Are there any proposed features or changes that could help make this problem go away over time?

**Comment by emilio** [comment]
See https://github.com/w3c/csswg-drafts/issues/1766. Not very easy to do them auto-magically work because it'd be cyclic (declarations using vw units could cause scrollbars to appear which would change how vw would resolve).

It'd be feasible to make them work with with overflow: scroll, but only Firefox implemented that and nobody else did so we removed that...

**Comment by simevidas** [comment]
@emilio I did not suggest to make changes to viewport units. I asked if the Working Group plans to do something about the problem of pointless horizontal scrollbars that appear on some websites, a problem that I predict will not go away on its own. Does this problem require an intervention or can it be ignored?

**Comment by emilio** [comment]
I guess we can reopen, but it's hard to say from a web engine when a scrollbar is "useless" if there's content overflowing.

**Comment by felipeerias** [comment]
Related: https://github.com/w3c/csswg-drafts/issues/5254

**Comment by felipeerias** [comment]
Also related: https://github.com/w3c/csswg-drafts/issues/5232

**Comment by coreyworrell** [comment]
> Not very easy to do them auto-magically work because it'd be cyclic (declarations using vw units could cause scrollbars to appear which would change how vw would resolve).

Percentages are not cyclic now, so why couldn't `vw` (or new unit) be equivalent to "percentage of root element"? Every single time I've used `vw` units, I don't care one little bit how wide the viewport is if part of it is being covered by a scrollbar. I only care what part of the viewport I can "draw" the element in without causing horizontal scrolling. If my element is direct child of the body, this is super easy, `100%` (or just don't specify any width is element is block).

```html
<body>
  <div class="one" style="width: 100%;"></div>
  <div class="two" style="width: 50%;">
    <div class="two-child" style="width: 100vw;"></div>
  </div>
</body>
```

`.one` and `.two-child` should resolve as same exact width regardless of vertical scrollbar being visible or not. Can we make this happen somehow?

**Comment by emilio** [comment]
> Percentages are not cyclic now, so why couldn't vw (or new unit) be equivalent to "percentage of root element"?

Because percentages aren't resolved until layout, while viewport units are. More generally, content inside the viewport can't affect the viewport size, but it can affect whether the root element shows scrollbars or not.

**Comment by coreyworrell** [comment]
@emilio Okay so why not create a new unit (or modify `vw`) that resolves at layout like percentage?

**Comment by emilio** [comment]
Well, that is one option. Note that that would behave subtly differently than the current viewport units. Adding layout-time relative units is relatively hard (no pun intended) compared to adding units that resolve at computed-value time like vh / vw. Though definitely not impossible.

**Comment by bramus** [comment]
Also related: https://github.com/w3c/csswg-drafts/issues/4329

**Comment by coreyworrell** [comment]
@emilio I don't know too much about the actual rendering processes, but [this description](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/render-tree-construction) makes it seem like it would be easier to calculate as there wouldn't need to be any tree traversal to compute, there would just be the one root node to be concerned with. But I understand there has to be more to it.

> Up to this point we've calculated which nodes should be visible and their computed styles, but we have not calculated their exact position and size within the viewport of the device---that's the "layout" stage, also known as "reflow."
>
> To figure out the exact size and position of each object on the page, the browser begins at the root of the render tree and traverses it.
> 
> The body of the above page contains two nested div's: the first (parent) div sets the display size of the node to 50% of the viewport width, and the second div---contained by the parent---sets its width to be 50% of its parent; that is, 25% of the viewport width.

Very confusing as their use of "viewport width" isn't the same as the `vw` unit, which you would reasonably assume.

**Comment by johannesodland** [comment]
Also related: https://github.com/w3c/csswg-drafts/issues/4691

**Comment by simevidas** [comment]
Related https://github.com/oddbird/sliiides/issues/3

**Comment by frivoal** [comment]
> It'd be feasible to make them work with with overflow: scroll, but only Firefox implemented that and nobody else did so we removed that...

I'm thinking we should go back to making them work with `overflow: scroll` and also with `scrollbar-gutter:stable` and `scrollbar-gutter:always`set on the root.

**Comment by hyperfekt** [comment]
Is my understanding correct that there is currently no way to size elements to the viewport minus the scrollbar while also having non-floating elements outside the viewport, since using `100%` to size them would give them a size larger than the viewport, and using `100vh`/`100vw`to size them will have part of them obscured by the scrollbar.?

**Comment by um83** [comment]
I suppose, that a new unit should be implemented, which determine viewport width and height without scrollbars. 
Without that it's not possible to make some cool stuff like horizontal document scrolling with sticky headers. 
Here's what I'm talking about: https://codepen.io/um83/pen/LYWBEmY

**Comment by brunoais** [comment]
> I suppose, that a new unit should be implemented, which determine viewport width and height without scrollbars.
> Without that it's not possible to make some cool stuff like horizontal document scrolling with sticky headers.
> Here's what I'm talking about: https://codepen.io/um83/pen/LYWBEmY

this? https://github.com/w3c/csswg-drafts/issues/6113

**Comment by um83** [comment]
> this? #6113

yep! exactly!

**Comment by jonjohnjohnson** [comment]
> I'm thinking we should go back to making them work with `overflow: scroll` and also with `scrollbar-gutter:stable` and `scrollbar-gutter:always`set on the root.

@frivoal Do you think you'd still rather go this way instead of having non-overlay scrollbars (both inline and block) affect the new `sv*` and/or `dv*` units? See https://github.com/w3c/csswg-drafts/issues/4329#issuecomment-863677668

**Comment by mor10** [comment]
To approach this from a "how does someone without in-depth knowledge of the functionings of the browser understand the concept of the viewport" perspective, I've talked to developers frustrated by this issue. The mental model boils down to this:

- When a scrollbar is fixed (as on Windows), the expectation is for the viewport width to be the width of the active canvas (so viewport minus scrollbar) because the scrollbar is understood as part of the chrome, not the viewport.
- When a scrollbar is dynamic (as on MacOS), the expectation is for the viewport width to be the width of the full viewport, and the dynamic scrollbar is seen as an overlay on top of the active canvas, so no change should happen to the width value when the scrollbar appears/disappears.

From this perspective, which is the user perspective in this context, the implementation of any new units like `dv*` should be as close to the _visually understood_ behavior as possible, ie if something looks to be part of the chrome, it is treated as part of the chrome, and if something looks to be an overlay, it is treated as an overlay.

**Comment by brunoais** [comment]
Maybe the distinguishing factor is whether the thing (scroll bar, in your example) is visible when the user is idling (I.e. not interacting with the page).
As a preliminary condition: If it disappears when interaction stops, it's non-counted overlay. If it stays (such as the browser toolbar in mobile browsers), it's counted overlay and taken into account with the `dv*` units.

What do you think?

**Comment by mor10** [comment]
@brunoais Yes, that's a different (and possibly clearer) way of saying the same thing.

**Comment by mirisuzanne** [comment]
Whatever we're able to do here should likely also apply to container (`cq*`) units. I like @mor10's summary of expected behavior here. It would also potentially be useful to have the ability to do something like `calc(100vw - scrollgutter)`.

**Comment by brunoais** [comment]
> Whatever we're able to do here should likely also apply to container (`cq*`) units. I like @mor10's summary of expected behavior here. It would also potentially be useful to have the ability to do something like `calc(100vw - scrollgutter)`.

[`100dvw`](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#relative_length_units) 😉 ?

**Comment by ahmadalfy** [comment]
> > Whatever we're able to do here should likely also apply to container (`cq*`) units. I like @mor10's summary of expected behavior here. It would also potentially be useful to have the ability to do something like `calc(100vw - scrollgutter)`.
> 
> [`100dvw`](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#relative_length_units) 😉 ?

It is still behaving the same. on Chrome (108) and Safari, 100vw is still the same as 100dvw when scrollbars exist.

**Comment by brunoais** [comment]
> > > Whatever we're able to do here should likely also apply to container (`cq*`) units. I like @mor10's summary of expected behavior here. It would also potentially be useful to have the ability to do something like `calc(100vw - scrollgutter)`.
> > 
> > 
> > [`100dvw`](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#relative_length_units) 😉 ?
> 
> It is still behaving the same. on Chrome (108) and Safari, 100vw is still the same as 100dvw when scrollbars exist.

Then it's a browser bug complain to its maker

**Comment by bramus** [comment]
> > It is still behaving the same. on Chrome (108) and Safari, 100vw is still the same as 100dvw when scrollbars exist.
> 
> Then it's a browser bug. Complain to its maker

It’s not a browser bug. All browsers are doing what the spec says here ([_“any scrollbars are assumed not to exist”_](https://drafts.csswg.org/css-values-3/#viewport-relative-lengths:~:text=However%2C%20any%20scrollbars%20are%20assumed%20not%20to%20exist)), hence this issue.

**Comment by fantasai** [comment]
To be clear, the spec says this because the browser vendors asked for it to match their implementations. It didn't originally say this.

**Comment by brunoais** [comment]
I specifically specified to consider the scrollbars in my proposal so this wouldn't happen!

**Comment by gilmoreorless** [comment]
So just to make sure I've understood this correctly...

1. Developers have been continually frustrated that there's no viewport unit that maps to the space remaining after scrollbars are added. This is because of the mental model summarised by @mor10 [in an earlier comment](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-881026696).
2. @brunoais puts together a well thought-out proposal in #6113 for dynamic viewport units. It includes an algorithm for resolving potential infinite layout cycles regarding scrollbars, specifically to address the problem in point 1.
3. The proposal is added to the CSS Values 4 spec. But browser vendors asked for the spec to keep the text "any scrollbars are assumed not to exist", so that it matched their implementations. Even though that line of text **completely defeats the purpose of the `dv*` units in the first place!**
4. A representative from a browser vendor effectively tells the original author of the proposal that the author doesn't know how the units are meant to work: https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1305449868
5. We all end up with **yet another** set of viewport units that don't solve this scrollbar problem, even though that's what they were meant to do.

I can't speak for @brunoais, but I know that if I was in their position I'd be bloody furious about this process. To this outsider it's really not a good look for the treatment of external contributions.

I come across sites and apps with unintended horizontal scrollbars like this almost daily. The most frustrating thing for me is that some browser vendors have previously said taking the scrollbars into account is close to impossible. Except Firefox has previously implemented that behaviour, so it's clearly possible to some extent. And as noted above, the original proposal outlined what to do to avoid infinite cycles.

**Comment by brunoais** [comment]
Thank you.
Thankfully, due to how I am, I'm not furious. However, I'm seriously disappointed. Feels like right to repair in USA's NY state.

**Comment by bramus** [comment]
> 4\. A representative from a browser vendor effectively tells the original author of the proposal that the author doesn't know how the units are meant to work: [[css-values] Use of 100vw is causing pointless horizontal scrollbars on some websites #6026 (comment)](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1305449868)

FWIW, I was trying to point out that – at this moment – it is not considered a browser bug, as the browsers follow what the spec says. You could say it’s a spec issue, though. Apologies if this came across differently.

**Comment by gilmoreorless** [comment]
> You could say it’s a spec issue, though. Apologies if this came across differently.

Fair enough, seems like it was my misinterpretation, apologies in return. I think it was the "hence this issue" that twisted my interpretation of the comment, as it seemed (to me) to be a circular argument regarding `dv*` units. I've edited my comment accordingly.

**Comment by simevidas** [comment]
Has anyone ever suggested a dynamic environment variable that would return the *current* viewport size reduction caused by a classic scrollbar on the page? Would that be possible? Would that make sense?

```css
.full-width {
  width: calc(100vw - env(viewport-reduction-inline));
}
```

So if the page does not have a vertical classic scrollbar, `env(viewport-reduction-inline)` returns `0px`, and the element is `100vw` wide, but if the page does have such a scrollbar, the environment variable returns how much inline space this scrollbar takes (usually 15 to 17 pixels), and the page can subtract that from `100vw` to get the actual viewport width and then set the element to that width.

This does not seem like a good solution to me, but I wanted to put it out there, since I haven’t seen this suggested yet.

**Comment by bramus** [comment]
> Has anyone ever suggested a dynamic environment variable that would return the _current_ viewport size reduction caused by a classic scrollbar on the page? Would that be possible? Would that make sense?

In https://github.com/web-platform-tests/interop-2022-viewport/issues/15#issuecomment-1268596477 this was discussed, and @emilio noted that you can’t have a dynamic value because it depends on layout and can cause layout cycles: _“Only reasonable thing to expose is the default size”_ … which brings us back to `env(scrollbar-inline-size)`.

**Comment by brunoais** [comment]
@simevidas https://github.com/w3c/csswg-drafts/issues/6113

**Comment by emilio** [comment]
I think exposing `scrollbar-inline-size` _plus_ a media query for overlay scrollbars is the most straight-forward solution. Those are things you can trivially determine via script anyways.

**Comment by brunoais** [comment]
> Those are things you can trivially determine via script anyways.

The idea is to solve layout problems with CSS, not with js. Js is for interactivity. Layout with Js is and will always be a hack

**Comment by kizu** [comment]
In our product we're currently essentially calculating the `scrollbar-inline-size` via JS, creating an invisible block with an overflow, and retrieving its width if it has one, then storing it as a custom property on root, to be used in calculations.

Being able to use a keyword for this, alongside a query to detect if scrollbars are present on a container would be very helpful.

Why I'm mentioning “keyword” and “container” rather than an “environment variable” and “media query”: we have `scrollbar-width` which can control if we have the scrollbar, and we can have a container that _can_ be scrolled programmatically when it has `overflow: hidden`. That means that authors could want to a) detect if a certain scrollport has a visible scrollbar, and b) get the width of a scrollbar in that case, which can be different for different containers (`scrollbar-width` and also current webkit scrollbar pseudo-elements).

**Comment by simevidas** [comment]
I don’t think `scrollbar-inline-size` is a solution to the `100vw` problem that’s being discussed in this issue. See my comment in https://github.com/w3c/csswg-drafts/issues/4691#issuecomment-1715038961 for why I think that is.

If CSS won’t answer the question “Is a vertical classic scrollbar currently present on the page?”, then I think the best way to deal with the `100vw` problem would be to reduce the size of `100vw` when `scrollbar-gutter: stable` or `overflow-y: scroll` is set on `<html>`, as proposed in https://github.com/w3c/csswg-drafts/issues/5254. With that change, `width: 100vw` would no longer cause horizontal overflow on pages with stable gutters. That seems to be the best we can do.

**Comment by simevidas** [comment]
As @kizu’s pointed out in https://github.com/w3c/csswg-drafts/issues/4691#issuecomment-1715388016, container query units can help with this problem. After turning `<body>`* into a query container, `100cqw` can be used as a substitute for `100vw` to get the desired behavior. Unlike `100vw`, `100cqw` becomes smaller when a classic scrollbar appears on the page. 

(*I use `<body>` because it didn’t work on `<html>` in Safari. I filed a [WebKit issue](https://bugs.webkit.org/show_bug.cgi?id=261649).)

For example: 

```css
body {
  margin: 0;
  container-type: inline-size;
}

.full-width {
  position: relative;
  left: calc(50% - 50cqw);
  width: 100cqw;
}
```

Demo: https://codepen.io/simevidas/pen/yLGzYBz?editors=0100

As far as I can see, `100cqw` behaves the way developers wanted `100vw` to behave with regards to classic scrollbars. In that case, I guess this issue can be closed. Developers can use `100cqw` instead of `100vw` for the “full-width” use-case, and in other situations where they need the actual viewport width in CSS.

**Comment by yuito-git** [comment]
Why can a custom property registered with @property save "page width"?
When I tried with "100%", it did not work.
```css
@property --page-width {
  syntax: '<length-percentage>';
  inherits: true;
  initial-value: 0px;
}

body {
  --page-width:100%;
}

p {
  width: var(--page-width); /*Width of parent element, not page width*/
}

```
Demo: https://codepen.io/mikiprogram/pen/QWzzvMR

Is there a statement on this in the following W3C page?
URL: https://www.w3.org/TR/css-properties-values-api-1/
I would be glad to know!

**Comment by simevidas** [comment]
@yuito-git Sorry, there was a mistake in my code. The `--page-width: 100cqw` declaration needs to be in a *child * of `<body>`. It doesn’t work if the declaration is in `<body>` itself. I’ve corrected my code and provided a second demo that shows that this solution works in browsers (Chrome and Safari, for now).

**Comment by yuito-git** [comment]
@simevidas Thank you very much! I found that the fix works for "100cqw". However, if I change the value of the CSS variable "page-width" from "100cqw" to "100%", I get the width of the parent element, not the page width. Do you know anything about this behavior?

**Comment by simevidas** [comment]
@yuito-git In my **demo 2**, if you just replace `--page-width: 100cqw;` with `--page-width: 100%;`, the value of `--page-width` will be `0px` because that custom property is registered as a `<length>`, so trying to assign a percentage results in the declaration being ignored by the browser. You can change the custom property’s type to `<length-percentage>`.

**Comment by yuito-git** [comment]
@simevidas In my demo, I have specified "<length-percentage>" for "syntax", but it is not the page width. Why is this?

**Comment by bfgeek** [comment]
This resolution might be a little scary for existing sites who are already working around this problem via something like:
`width: calc(100vw - var(--scrollbar-width))` was this discussed?

**Comment by bfgeek** [comment]
E.g. https://github.com/search?q=%22calc%28100vw+-+16px%22&type=code

**Comment by emilio** [comment]
@bfgeek those would be less wide than needed, but that seems fine to me, it's already what happens in platforms with overlay scrollbars, right?

**Comment by tabatkins** [comment]
@brunoais 
> Why not move that resolution to 100 dvw & 100 dvh

The resolution applies to all of the viewport units.

**Comment by bfgeek** [comment]
> @bfgeek those would be less wide than needed, but that seems fine to me, it's already what happens in platforms with overlay scrollbars, right?

@emilio Lots of enterprise like apps use ::-webkit-scrollbar to force a scrollbar to exist.

**Comment by kizu** [comment]
> This resolution might be a little scary for existing sites who are already working around this problem via something like:
width: calc(100vw - var(--scrollbar-width))

I might confirm that in our product we have at least 12 places where we have almost exactly this: we detect the current scrollbar width, apply it as a variable, and then subtract when necessary from the `100vw`.

That said, we're not using `overflow: scroll` or `scrollbar-gutter: stable` on the root scroller, so for us this _should be_ safe. And, overall, that requirement would probably reduce the potential impact.

Then, the impact itself would _usually_ not be very drastic, in most cases resulting in something being a bit narrower than necessary. Still a concern, though.

**Comment by tabatkins** [comment]
Yeah, scrollbar-gutter has relatively low usage right now, as does overflow:scroll on the root (as opposed to auto). And the failure mode is "things are slightly narrower than they should be", which is generally pretty okay (as opposed to the current failure mode of people naively using 100vw and testing on overlay-scrollbar browsers, which is "horizontal scrollbars on everything that just barely wiggle back and forth", a much more annoying failure mode).

**Comment by bramus** [comment]
Some situation that came to mind today:

1. Say the window’s `innerWidth` is 1000px. After subtracting the scrollbar-width _(assuming a classic `16px`)_ that would have `100vw` be computed to `984px`.
2. Now say on the page you have a scrollable `<div>` with a width of `50vw`. Taking half of `984px`, that would give it a width of `492px`. Because of its scrollbar, there’s “only” `476px` available space inside of it left though.
3. Now comes the part where authors might get tripped up: if they then size something inside that `<div>` to be `50vw`, they will end up with a horizontal scrollbar again, as the computed width is `492px` and not the leftover `476px`.

If you know what’s happening that is expected behavior, yet it might trip up some authors. It also underlines that this automatic subtraction only solves the `100vw` case on the root scroller. A solution with an envvar would allow fixing all those cases, but puts the burden on the author to take it into account.

Related to the third bullet point above: in that case, maybe we should also do the same for `100cqi` with the same conditions on the container? Demo: https://codepen.io/bramus/pen/rNPqOZe

**Comment by brunoais** [comment]
> Some situation that came to mind today:

That's part of why I think this should be applied to `dvh` and `dvw`. I think we shouldn't break such year long used variables and let's just break the ones that were already [designed from the ground up to work like that.](https://github.com/w3c/csswg-drafts/issues/6113)

**Comment by simevidas** [comment]
> maybe we should also do the same for 100cqi with the same conditions on the container?

This kind-of makes it look like it’s not a good idea to turn scroll containers into  inline-size query containers in general. We might be seeing a repeat of the `100vw` problem in the future where developers test their CSS only in browsers with overlay scrollbars, and they’re not aware that their container query units are causing horizontal scrollbars.

**Comment by markcellus** [comment]
> I think this should be applied to `dvh` and `dvw`.

Agree. It would be suprising if not. In fact, I was already assuming in other talks that everywhere we were saying "vw", it was assumed to also apply to "dvw". :laughing:

**Comment by brunoais** [comment]
@markcellus I wasn't explicit enough, sorry. I fixed it.

**Comment by tabatkins** [comment]
> Now comes the part where authors might get tripped up: if they then size something inside that <div> to be 50vw, they will end up with a horizontal scrollbar again, as the computed width is 492px and not the leftover 476px.

I'm confused. That would happen *regardless* - if the 50vw-wide div has scrollbars, its inner width is going to be less than 50vw no matter how we define vw.

> A solution with an envvar would allow fixing all those cases, but puts the burden on the author to take it into account.

Also not sure what you mean here. How would an env() make this case work?

**Comment by astearns** [comment]
I removed the Agenda+ tag for now because the debate here seems ongoing.

**Comment by bfgeek** [comment]
@astearns  I think we still want to undo the previous resolution to reconsider

**Comment by astearns** [comment]
I have not yet seen that you have convinced anyone else about this. From my personal perspective some loss of usable width in some (fixable?) cases is preferable to an unwanted horizontal scrollbar

**Comment by bfgeek** [comment]
> I have not yet seen that you have convinced anyone else about this. 

I think I have?

**Comment by brunoais** [comment]
I request to consider not adding the latest resolution to 100vw and 100vh but all the other *vw and *vh is OK.

**Comment by simevidas** [comment]
> I request to consider not adding the latest resolution to 100vw and 100vh but all the other *vw and *vh is OK.

@brunoais What’s the risk here? Are there websites with `overflow: scroll` on `<html>` that *don’t* want `100vw` to shrink by a very small amount? What exactly could break?

**Comment by mdmoreau** [comment]
Is it possible to create a property similar to box-sizing (viewport-sizing?) that only works on :root? That way people could explicitly opt in to the behavior, similar to the way most sites use box-sizing: border-box instead of the default content-box.

**Comment by bramus** [comment]
> > Now comes the part where authors might get tripped up: if they then size something inside that  to be 50vw, they will end up with a horizontal scrollbar again, as the computed width is 492px and not the leftover 476px.
> 
> ([#](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1834696746)) I'm confused. That would happen _regardless_ - if the 50vw-wide div has scrollbars, its inner width is going to be less than 50vw no matter how we define vw.

I know that will happen regardless, but it might trip up authors as `vw` on root does account for the root scrollbar but when used elsewhere it does not take the (direct) parent scroller’s scrollbar into account.

That confusion could be solved by doing something similar for `cqi`/`cqw` under the same conditions [which I also suggested](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1833818113)) and @mirisuzanne [also hinted at this before](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1163438639).

_(This looks like food for a follow-up issue)_

> > A solution with an envvar would allow fixing all those cases, but puts the burden on the author to take it into account.
> 
> ([#](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1834696746)) Also not sure what you mean here. How would an env() make this case work?

The `env()` proposal would allow authors to manually take the size of the scrollbar into account. It is a fixed value that represents the default scrollbar size. This would work for both the viewport and containers.

```
width: calc(100vw) - env(scrollbar-inline-size);
width: calc(100cqw) - env(scrollbar-inline-size);
```

This mimics what [authors are currently doing](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1832490269), with the different that the scrollbar-width is reported by the browser here.

But, as @emilio [mentioned](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1713351193), this would also require a way to detect classic scrollbars. It would also require a way to detect if something is scrolling or not … which would make this all very complicated for authors to implement, as demonstrated in [this snippet](https://gist.github.com/bramus/bcca5788d8ced82837180b7a15760c84) I presented at the 2023 summer F2F.

So yeah, maybe this `envvar()` route isn’t the preferable path indeed … too much jumping through hoops.

---

Winging back to @bfgeek’s [concern](https://github.com/w3c/csswg-drafts/issues/6026#issuecomment-1832490269): I’ve asked some colleagues to query the HTTP archive to see:

1. How often the `width: calc(100vw - var(--scrollbar-width))` pattern is used.
2. How often `overflow: scroll` or `scrollbar-gutter: stable` on the `:root` is used.
3. How often both are used together

I’m still waiting on the results of these queries.

**Comment by coreyworrell** [comment]
Some use cases: most of our sites uses `overflow: scroll` on the root to avoid layout shift between pages (first page does not cause overflow, next page does) or between UI states (toggling a details summary item causes scrollbar to appear).

These sites often have modal UI that hides overflow on root (`scroll-behavior: contain` will not work if modal itself is not large enough to cause scrolling, and only applies if pointer is inside the modal anyways), and extra padding equal to the scrollbar width must be applied to avoid layout shift when showing/hiding the modal (as the scollbar hides and shows respectively).

Dropdown menus with absolutely positioned children (ie: full width dropdown that "breaks out" of it's parent), rely on using the `width: calc(100vw - var(--scrollbar-width))` pattern.

**Comment by fantasai** [comment]
Filed follow-up wrt `dv*` etc in https://github.com/w3c/csswg-drafts/issues/10059
